### ISSUE TITLE
💅 false positive of `noDelete` with `process.env`

### ISSUE THREAD

**Issue by BenjaminBrienen** [biomejs/biome#11014]
### Environment information

<details>

```bash

CLI:
  Version:                      2.5.4
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_DISTRIBUTION:                npm
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.17.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/11.13.1

Biome Configuration:
  Status:                       Not set
  Path:                         unset

Workspace:
  Open Documents:               0

```

</details>

### Rule name

noDelete

### Playground link

https://biomejs.dev/playground/?lintRules=noDelete&tab=formatter&pane=Diagnostics&code=ZABlAGwAZQB0AGUAIABwAHIAbwBjAGUAcwBzAC4AZQBuAHYAWwAiAFQARQBTAFQAXwBWAEEAUgBJAEEAQgBMAEUAIgBdAA%3D%3D

### Expected result

With the code:
```
suite.addSyncTest("Should support external variables", () => {
	process.env["TEST_VARIABLE"] = "test";
	const envJson = {
		USING_EXTERNAL_VAR: "${env:TEST_VARIABLE} test ${env:TEST_VARIABLE}",
	};
	const expectedEnv = {
		USING_EXTERNAL_VAR: "test test test",
	};

	const actualEnv = substituteVariablesInEnv(envJson);
	deepStrictEqual(actualEnv, expectedEnv);
	// biome-ignore lint/complexity/useLiteralKeys: Property 'TEST_VARIABLE' comes from an index signature
	delete process.env["TEST_VARIABLE"];
});
```
I get `noDelete` triggering on the `delete process.env["TEST_VARIABLE"];` even though it is a sort of map and is technically correct as far as I can tell.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct
